### PR TITLE
crates refactors

### LINKED ISSUE #2111: Restructure codebase for a module-based separation of concerns
At the moment, we split the codebase into separate directories and sub-packages based on technical aspects:
- `apps/` includes CLI code, wallet code, and a bunch of ABCI handling code
- `core/` includes types, storage layouts, and some miscellaneous state machine logic
- `ethereum-bridge/` contains a bunch of Ethereum bridge-related code
- `proof-of-stake/` contains a bunch of proof-of-stake related code
- `sdk/` contains the SDK code (client code)
- `shared/` contains validity predicate logic, miscellaneous types and events logic, some protocol structure, and some WASM VP wrapper functions
- `tx_prelude`, `vp_prelude`, and `vm_env` contain some WASM standard library code (as far as I can tell)

At least personally, I find this separation somewhat confusing, for a few reasons:
- it's difficult to find all of the code associated with a particular feature (e.g. governance), since it's split into many places
- it's not always entirely clear whether some piece of code is in the right place, and some locations seem inconsistent
- I think it probably creates more file conflicts than necessary, since code for unrelated features is not always clearly abstracted

I propose for discussion a rework of the structure of the codebase focused on _feature modules_, e.g.:
- A _module_ is a defined interface which must implement a bunch of functions, such as:
    - a validity predicate
    - logic to be run at the beginning/end of blocks/epochs
    - transaction builder logic for the sdk
    - commands to expose in the cli
    - genesis fields required
- Governance/PGF, proof-of-stake, IBC, the Ethereum bridge, MASP all become _modules_
    - if convenient, these can be implemented as separate crates in some `modules/` directory   
    - all of these modules should implement this interface
    - no feature-specific code should live outside its module

Some precedent for this abstraction can be found in the [modules system of the Cosmos SDK](https://docs.cosmos.network/v0.46/building-modules/intro.html), which, although it has a lot of boilerplate, implements this basic module abstraction in a reasonably clean manner.

I think that if we do this well, it could make the codebase much easier to navigate, reduce file conflicts, and make it easier to add new modules in the future. Commentary & discussion welcome.

The dependency graph from 0.31.8:
deps
(generated with `cargo depgraph --workspace-only --dedup-transitive-deps | dot -Tpng `)

Module dependencies that should be removed:
- trans_token <-
	- [ ] governance
- parameters <-
	- [ ] governance
	- [ ] shielded_token
- account <-
	- [ ] proof_of_stake
- governance <-
	- [ ] proof_of_stake
	- [ ] ibc
- token <-
	- [ ] ibc
- proof_of_stake <-
	- [ ] ethereum_bridge

### PR DESCRIPTION
**Pull request by tzemanovic**
## Describe your changes 

closes #2111

## Checklist before merging 
- [ ] If this PR has some consensus breaking changes, I added the corresponding `breaking::` labels
    - This will require 2 reviewers to approve the changes

### CODE REVIEW

**Comment by Fraccaman** [comment]
im not sure but do u mind checking if some of the paths in `codecov.yml` have changed?

**Comment by tzemanovic** [comment]
> im not sure but do u mind checking if some of the paths in `codecov.yml` have changed?

not that I can see, why?

File: crates/node/src/protocol.rs [hunk 1]
```diff
@@ -1185,232 +1159,233 @@ where
...
+                            InternalAddress::TempStorage => Err(
+                                // Temp storage changes must never be committed
+                                Error::AccessForbidden(
+                                    (*internal_addr).clone(),
+                                ),
+                            ),
+                            InternalAddress::ReplayProtection => Err(
+                                // Replay protection entries should never be
+                                // written to
+                                // via transactions
```
Comment (sug0) [inline crates/node/src/protocol.rs:1343]: nit
```suggestion
                                // Replay protection entries should never be
                                // written to via transactions
```

**Review by sug0 — CHANGES_REQUESTED** [review]

File: crates/core/src/chain.rs [hunk 2]
```diff
@@ -285,6 +113,432 @@ impl ChainId {
...
+    /// Timestamp associated to block
+    pub time: DateTimeUtc,
+    /// Hash of the addresses of the next validator set
+    pub next_validators_hash: Hash,
+}
+
+impl BlockHeader {
+    /// The number of bytes when this header is encoded
+    pub fn encoded_len(&self) -> usize {
+        self.serialize_to_vec().len()
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply crates/core/src/chain.rs]: I replaced it with a const fn in [c71daae](https://github.com/anoma/namada/pull/3670/commits/c71daae2ae1f343a81d1e91ccfaa1a826afa64b5)

File: crates/core/src/chain.rs [hunk 3]
```diff
@@ -285,6 +113,432 @@ impl ChainId {
...
+}
+
+#[allow(missing_docs)]
+#[derive(Error, Debug)]
+pub enum ParseBlockHashError {
+    #[error("Error parsing block hash: {0}")]
+    ParseBlockHash(String),
+}
+
+impl TryFrom<Vec<u8>> for BlockHash {
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply crates/core/src/chain.rs]: done in [c71daae](https://github.com/anoma/namada/pull/3670/commits/c71daae2ae1f343a81d1e91ccfaa1a826afa64b5) (it was already there so just rm'd this one)

**Review by sug0 — APPROVED** [review]

**Pull request merged**